p2p/server/deadline_adjuster.go

```diff
@@ -0,0 +1,130 @@
+package server
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/jonboulle/clockwork"
+	"github.com/libp2p/go-yamux/v4"
+)
+
+const (
+	deadlineAdjusterChunkSize = 4096
+)
+
+type deadlineAdjuster struct {
+	peerStream
+	desc            string
+	timeout         time.Duration
+	hardTimeout     time.Duration
+	totalRead       int
+	totalWritten    int
+	start           time.Time
+	clock           clockwork.Clock
+	chunkSize       int
+	nextAdjustRead  int
+	nextAdjustWrite int
+	hardDeadline    time.Time
+}
+
+func newDeadlineAdjuster(stream peerStream, desc string, timeout, hardTimeout time.Duration) *deadlineAdjuster {
+	return &deadlineAdjuster{
+		peerStream:      stream,
+		desc:            desc,
+		timeout:         timeout,
+		hardTimeout:     hardTimeout,
+		start:           time.Now(),
+		clock:           clockwork.NewRealClock(),
+		chunkSize:       deadlineAdjusterChunkSize,
+		nextAdjustRead:  -1,
+		nextAdjustWrite: -1,
+	}
+}
+
+func (dadj *deadlineAdjuster) augmentError(what string, err error) error {
+	if !errors.Is(err, context.DeadlineExceeded) && !errors.Is(err, yamux.ErrTimeout) {
+		return err
+	}
+	now := dadj.clock.Now()
+	return fmt.Errorf("%s: %s after %v, %d bytes read, %d bytes written, timeout %v, hard timeout %v: %w",
+		dadj.desc,
+		what,
+		now.Sub(dadj.start),
+		dadj.totalRead,
+		dadj.totalWritten,
+		dadj.timeout,
+		dadj.hardTimeout,
+		err)
+}
+
+func (dadj *deadlineAdjuster) adjust() error {
+	now := dadj.clock.Now()
+	if dadj.hardDeadline.IsZero() {
+		dadj.hardDeadline = now.Add(dadj.hardTimeout)
+	} else if !now.Before(dadj.hardDeadline) {
```

Comment:
NIT: It's simpler to avoid negation if possible.
```suggestion
	} else if now.After(dadj.hardDeadline) {
```

Reply:
well it is not quite the same (`!now.Before(...)` includes equal times), but given the resolution of `time.Time` this doesn't matter, so ok :)